Bug in `ConvTranspose`?

The `Lux.ConvTranspose` layer does not throw an error when the number of input channels is **larger** than the `in_chs` specified in the layer. If the number of input channels is smaller, it correctly throws an error.


See this minimal example:

First, setup the layer with `42` `in_chs` and `16` `out_chs`:

```julia
julia> using Lux, Random
julia> ct = ConvTranspose((4, 4), 42=>16, stride=2, pad=1)
ConvTranspose((4, 4), 42 => 16, pad=1, stride=2)  # 10_768 parameters

julia> ps, st = Lux.setup(Xoshiro(0), ct);
```

---

Now, with the correct number of channels, it's of course correct...

```julia
julia> ct_out, ct_st = ct(randn(Float32, 28, 28, 42, 8), ps, st);

julia> size(ct_out)
(56, 56, 16, 8)
```

... and with a too-small number of input channels, it throws an error, as expected:

```julia
julia> ct_out, ct_st = ct(randn(Float32, 28, 28, 23, 8), ps, st);
ERROR: TaskFailedException

    nested task error: AssertionError: DimensionMismatch("Data output channel count (23 vs. 42)")
    Stacktrace:
     [1] check_dims(x::NTuple{5, Int64}, w::NTuple{5, Int64}, y::NTuple{5, Int64}, cdims::DenseConvDims{3, 3, 3, 6, 3})
       @ NNlib ~/.julia/packages/NNlib/5iRSB/src/dim_helpers/DenseConvDims.jl:81
     [2] ∇conv_data_im2col!(dx::SubArray{Float32, 5, Array{Float32, 5}, Tuple{Base.Slice{Base.OneTo{Int64}}, Base.Slice{Base.OneTo{Int64}}, Base.Slice{Base.OneTo{Int64}}, UnitRange{Int64}, Base.Slice{Base.OneTo{Int64}}}, false}, dy::SubArray{Float32, 5, Array{Float32, 5}, Tuple{Base.Slice{Base.OneTo{Int64}}, Base.Slice{Base.OneTo{Int64}}, Base.Slice{Base.OneTo{Int64}}, UnitRange{Int64}, Base.Slice{Base.OneTo{Int64}}}, false}, w::SubArray{Float32, 5, Array{Float32, 5}, Tuple{Base.Slice{Base.OneTo{Int64}}, Base.Slice{Base.OneTo{Int64}}, Base.Slice{Base.OneTo{Int64}}, Base.Slice{Base.OneTo{Int64}}, UnitRange{Int64}}, true}, cdims::DenseConvDims{3, 3, 3, 6, 3}; col::Array{Float32, 3}, alpha::Float32, beta::Float32, ntasks::Int64)
       @ NNlib ~/.julia/packages/NNlib/5iRSB/src/impl/conv_im2col.jl:132
     [3] ∇conv_data_im2col!(dx::SubArray{Float32, 5, Array{Float32, 5}, Tuple{Base.Slice{Base.OneTo{Int64}}, Base.Slice{Base.OneTo{Int64}}, Base.Slice{Base.OneTo{Int64}}, UnitRange{Int64}, Base.Slice{Base.OneTo{Int64}}}, false}, dy::SubArray{Float32, 5, Array{Float32, 5}, Tuple{Base.Slice{Base.OneTo{Int64}}, Base.Slice{Base.OneTo{Int64}}, Base.Slice{Base.OneTo{Int64}}, UnitRange{Int64}, Base.Slice{Base.OneTo{Int64}}}, false}, w::SubArray{Float32, 5, Array{Float32, 5}, Tuple{Base.Slice{Base.OneTo{Int64}}, Base.Slice{Base.OneTo{Int64}}, Base.Slice{Base.OneTo{Int64}}, Base.Slice{Base.OneTo{Int64}}, UnitRange{Int64}}, true}, cdims::DenseConvDims{3, 3, 3, 6, 3})
       @ NNlib ~/.julia/packages/NNlib/5iRSB/src/impl/conv_im2col.jl:126
     [4] (::NNlib.var"#323#327"{Base.Pairs{Symbol, Union{}, Tuple{}, NamedTuple{(), Tuple{}}}, DenseConvDims{3, 3, 3, 6, 3}, SubArray{Float32, 5, Array{Float32, 5}, Tuple{Base.Slice{Base.OneTo{Int64}}, Base.Slice{Base.OneTo{Int64}}, Base.Slice{Base.OneTo{Int64}}, Base.Slice{Base.OneTo{Int64}}, UnitRange{Int64}}, true}, SubArray{Float32, 5, Array{Float32, 5}, Tuple{Base.Slice{Base.OneTo{Int64}}, Base.Slice{Base.OneTo{Int64}}, Base.Slice{Base.OneTo{Int64}}, UnitRange{Int64}, Base.Slice{Base.OneTo{Int64}}}, false}, SubArray{Float32, 5, Array{Float32, 5}, Tuple{Base.Slice{Base.OneTo{Int64}}, Base.Slice{Base.OneTo{Int64}}, Base.Slice{Base.OneTo{Int64}}, UnitRange{Int64}, Base.Slice{Base.OneTo{Int64}}}, false}})()
       @ NNlib ./threadingconstructs.jl:410
Stacktrace: [...]
```

**However**, when providing too many channels, instead of failing:

```julia
julia> ct_out, ct_st = ct(randn(Float32, 28, 28, 66, 8), ps, st);

julia> size(ct_out)
(56, 56, 16, 8)
```

--- a/Project.toml
+++ b/Project.toml
@@ -1,7 +1,7 @@
 name = "Lux"
 uuid = "b2108857-7c20-44ae-9111-449ecde12c47"
 authors = ["Avik Pal <avikpal@mit.edu> and contributors"]
-version = "0.5.20"
+version = "0.5.21"
 
 [deps]
 ADTypes = "47edcb42-4c32-4615-8424-f2b9edc5f35b"

--- a/examples/PolynomialFitting/main.jl
+++ b/examples/PolynomialFitting/main.jl
@@ -94,6 +94,7 @@ dev_gpu = gpu_device()
 
 tstate = main(tstate, vjp_rule, (x, y), 250)
 y_pred = dev_cpu(Lux.apply(tstate.model, dev_gpu(x), tstate.parameters, tstate.states)[1])
+nothing #hide
 
 # Let's plot the results
 

--- a/examples/SimpleRNN/main.jl
+++ b/examples/SimpleRNN/main.jl
@@ -175,6 +175,7 @@ function main()
 end
 
 ps_trained, st_trained = main()
+nothing #hide
 
 # ## Saving the Model
 

--- a/src/layers/recurrent.jl
+++ b/src/layers/recurrent.jl
@@ -1,5 +1,18 @@
 abstract type AbstractRecurrentCell{use_bias, train_state} <: AbstractExplicitLayer end
 
+# Fallback for vector inputs
+function (rnn::AbstractRecurrentCell)(x::AbstractVector, ps, st::NamedTuple)
+    (y, carry), st_ = rnn(reshape(x, :, 1), ps, st)
+    return (vec(y), vec.(carry)), st_
+end
+
+function (rnn::AbstractRecurrentCell)((x, carry), ps, st::NamedTuple)
+    x_ = reshape(x, :, 1)
+    carry_ = map(Base.Fix2(reshape, (:, 1)), carry)
+    (y, carry_new), st_ = rnn((x_, carry_), ps, st)
+    return (vec(y), vec.(carry_new)), st_
+end
+
 abstract type AbstractTimeSeriesDataBatchOrdering end
 
 struct TimeLastIndex <: AbstractTimeSeriesDataBatchOrdering end
@@ -54,6 +67,24 @@ rather allows operating on an entire sequence of inputs at once. See
 ## States
 
   - Same as `cell`.
+
+:::tip
+
+Frameworks like Tensorflow have special implementation of
+[`MultiRNNCell`](https://www.tensorflow.org/api_docs/python/tf/compat/v1/nn/rnn_cell/MultiRNNCell)
+to handle sequentially composed RNN Cells. In Lux, one can simple stack multiple
+`Recurrence` blocks in a `Chain` to achieve the same.
+
+    Chain(
+        Recurrence(RNNCell(inputsize => latentsize); return_sequence=true),
+        Recurrence(RNNCell(latentsize => latentsize); return_sequence=true),
+        :
+        x -> stack(x; dims=2)
+    )
+
+For some discussion on this topic, see https://github.com/LuxDL/Lux.jl/issues/472.
+
+:::
 """
 struct Recurrence{
     R, C <: AbstractRecurrentCell, O <: AbstractTimeSeriesDataBatchOrdering} <:
@@ -69,6 +100,11 @@ end
 
 _eachslice(x::AbstractArray, ::TimeLastIndex) = _eachslice(x, Val(ndims(x)))
 _eachslice(x::AbstractArray, ::BatchLastIndex) = _eachslice(x, Val(ndims(x) - 1))
+function _eachslice(x::AbstractMatrix, ::BatchLastIndex)
+    error("`BatchLastIndex` not supported for AbstractMatrix. You probably want to use \
+           `TimeLastIndex`.")
+    return
+end
 
 @inline function (r::Recurrence)(x::AbstractArray, ps, st::NamedTuple)
     return Lux.apply(r, _eachslice(x, r.ordering), ps, st)

--- a/src/utils.jl
+++ b/src/utils.jl
@@ -65,6 +65,10 @@ get_typename(::T) where {T} = Base.typename(T).wrapper
 @inline _gate(x::AbstractMatrix, h::Int, n::Int) = view(x, _gate(h, n), :)
 
 @inline function _init_hidden_state(rng::AbstractRNG, rnn, x::AbstractMatrix)
+    return rnn.init_state(rng, rnn.out_dims, size(x, 2))
+end
+
+@inline function _init_hidden_state(rng::AbstractRNG, rnn, x::GPUArraysCore.AnyGPUMatrix)
     return convert(ArrayInterface.parameterless_type(parent(x)),
         rnn.init_state(rng, rnn.out_dims, size(x, 2)))
 end

--- a/src/utils.jl
+++ b/src/utils.jl
@@ -155,9 +155,14 @@ function _conv_transpose_dims(
     I = (size(x)[1:(end - 2)] .- 1) .* stride .+ 1 .+
         (size(weight)[1:(end - 2)] .- 1) .* dilation .- combined_pad
     C_in = size(weight)[end - 1] * groups
+    C_out = size(weight)[end]
     batch_size = size(x)[end]
-    # Create DenseConvDims() that looks like the corresponding conv()
     w_size = size(weight)
+    if size(x)[end - 1] != C_out
+        throw(DimensionMismatch("Expected $(C_out) input channels but got \
+                                 $(size(x)[end - 1]) channels."))
+    end
+    # Create DenseConvDims() that looks like the corresponding conv()
     return DenseConvDims(
         (I..., C_in, batch_size), w_size; stride, padding, dilation, groups)
 end

--- a/test/layers/conv_tests.jl
+++ b/test/layers/conv_tests.jl
@@ -660,5 +660,22 @@ end
 
         @test occursin("groups=2", sprint(show, ConvTranspose((3, 3), 2 => 4; groups=2)))
         @test occursin("2 => 4", sprint(show, ConvTranspose((3, 3), 2 => 4; groups=2)))
+
+        @testset "Catch Channel Mismatch Early: LuxDL/Lux.jl#455" begin
+            layer = ConvTranspose((4, 4), 42 => 16; stride=2, pad=1)
+
+            x = randn(Float32, 28, 28, 42, 3) |> aType
+            ps, st = Lux.setup(rng, layer) .|> device
+
+            @test_nowarn layer(x, ps, st)
+
+            x = randn(Float32, 28, 28, 46, 3) |> aType
+
+            @test_throws DimensionMismatch layer(x, ps, st)
+
+            x = randn(Float32, 28, 28, 23, 3) |> aType
+
+            @test_throws DimensionMismatch layer(x, ps, st)
+        end
     end
 end

--- a/test/layers/recurrent_tests.jl
+++ b/test/layers/recurrent_tests.jl
@@ -9,23 +9,25 @@
             RNNCell(3 => 5, identity; use_bias=false, train_state=false))
             __display(rnncell)
             ps, st = Lux.setup(rng, rnncell) .|> device
-            x = randn(rng, Float32, 3, 2) |> aType
-            (y, carry), st_ = Lux.apply(rnncell, x, ps, st)
+            for x_size in ((3, 2), (3,))
+                x = randn(rng, Float32, x_size...) |> aType
+                (y, carry), st_ = Lux.apply(rnncell, x, ps, st)
 
-            @jet rnncell(x, ps, st)
-            @jet rnncell((x, carry), ps, st)
+                @jet rnncell(x, ps, st)
+                @jet rnncell((x, carry), ps, st)
 
-            function loss_loop_rnncell(p)
-                (y, carry), st_ = rnncell(x, p, st)
-                for i in 1:10
-                    (y, carry), st_ = rnncell((x, carry), p, st_)
+                function loss_loop_rnncell(p)
+                    (y, carry), st_ = rnncell(x, p, st)
+                    for i in 1:10
+                        (y, carry), st_ = rnncell((x, carry), p, st_)
+                    end
+                    return sum(abs2, y)
                 end
-                return sum(abs2, y)
+
+                @test_throws ErrorException ps.train_state
+
+                @eval @test_gradients $loss_loop_rnncell $ps atol=1.0f-3 rtol=1.0f-3 gpu_testing=$ongpu
             end
-
-            @test_throws ErrorException ps.train_state
-
-            @eval @test_gradients $loss_loop_rnncell $ps atol=1.0f-3 rtol=1.0f-3 gpu_testing=$ongpu
         end
 
         @testset "Trainable hidden states" begin
@@ -33,19 +35,24 @@
                 RNNCell(3 => 5, identity; use_bias=true, train_state=true))
                 rnn_no_trainable_state = RNNCell(
                     3 => 5, identity; use_bias=false, train_state=false)
-                x = randn(rng, Float32, 3, 2) |> aType
                 _ps, _st = Lux.setup(rng, rnn_no_trainable_state) .|> device
-                (_y, _carry), _ = Lux.apply(rnn_no_trainable_state, x, _ps, _st)
 
                 rnncell = RNNCell(3 => 5, identity; use_bias=false, train_state=true)
                 ps, st = Lux.setup(rng, rnncell) .|> device
                 ps = merge(_ps, (hidden_state=ps.hidden_state,))
-                (y, carry), _ = Lux.apply(rnncell, x, ps, st)
-                @test carry == _carry
 
-                l, back = Zygote.pullback(p -> sum(abs2, 0 .- rnncell(x, p, st)[1][1]), ps)
-                gs = back(one(l))[1]
-                @test !isnothing(gs.hidden_state)
+                for x_size in ((3, 2), (3,))
+                    x = randn(rng, Float32, x_size...) |> aType
+                    (_y, _carry), _ = Lux.apply(rnn_no_trainable_state, x, _ps, _st)
+
+                    (y, carry), _ = Lux.apply(rnncell, x, ps, st)
+                    @test carry == _carry
+
+                    l, back = Zygote.pullback(
+                        p -> sum(abs2, 0 .- rnncell(x, p, st)[1][1]), ps)
+                    gs = back(one(l))[1]
+                    @test !isnothing(gs.hidden_state)
+                end
             end
         end
     end
@@ -59,85 +66,100 @@ end
             LSTMCell(3 => 5; use_bias=false))
             __display(lstmcell)
             ps, st = Lux.setup(rng, lstmcell) .|> device
-            x = randn(rng, Float32, 3, 2) |> aType
-            (y, carry), st_ = Lux.apply(lstmcell, x, ps, st)
 
-            @jet lstmcell(x, ps, st)
-            @jet lstmcell((x, carry), ps, st)
+            for x_size in ((3, 2), (3,))
+                x = randn(rng, Float32, x_size...) |> aType
+                (y, carry), st_ = Lux.apply(lstmcell, x, ps, st)
 
-            function loss_loop_lstmcell(p)
-                (y, carry), st_ = lstmcell(x, p, st)
-                for i in 1:10
-                    (y, carry), st_ = lstmcell((x, carry), p, st_)
+                @jet lstmcell(x, ps, st)
+                @jet lstmcell((x, carry), ps, st)
+
+                function loss_loop_lstmcell(p)
+                    (y, carry), st_ = lstmcell(x, p, st)
+                    for i in 1:10
+                        (y, carry), st_ = lstmcell((x, carry), p, st_)
+                    end
+                    return sum(abs2, y)
                 end
-                return sum(abs2, y)
+
+                @eval @test_gradients $loss_loop_lstmcell $ps atol=1.0f-3 rtol=1.0f-3 gpu_testing=$ongpu
+
+                @test_throws ErrorException ps.train_state
+                @test_throws ErrorException ps.train_memory
             end
-
-            @eval @test_gradients $loss_loop_lstmcell $ps atol=1.0f-3 rtol=1.0f-3 gpu_testing=$ongpu
-
-            @test_throws ErrorException ps.train_state
-            @test_throws ErrorException ps.train_memory
         end
 
         @testset "Trainable hidden states" begin
-            x = randn(rng, Float32, 3, 2) |> aType
-            _lstm = LSTMCell(3 => 5; use_bias=false, train_state=false, train_memory=false)
-            _ps, _st = Lux.setup(rng, _lstm) .|> device
-            (_y, _carry), _ = Lux.apply(_lstm, x, _ps, _st)
+            for x_size in ((3, 2), (3,))
+                x = randn(rng, Float32, x_size...) |> aType
+                _lstm = LSTMCell(
+                    3 => 5; use_bias=false, train_state=false, train_memory=false)
+                _ps, _st = Lux.setup(rng, _lstm) .|> device
+                (_y, _carry), _ = Lux.apply(_lstm, x, _ps, _st)
 
-            lstm = LSTMCell(3 => 5; use_bias=false, train_state=false, train_memory=false)
-            ps, st = Lux.setup(rng, lstm) .|> device
-            ps = _ps
-            (y, carry), _ = Lux.apply(lstm, x, ps, st)
-            @test carry == _carry
-            l, back = Zygote.pullback(p -> sum(abs2, 0 .- sum(lstm(x, p, st)[1][1])), ps)
-            gs = back(one(l))[1]
-            @test_throws ErrorException gs.bias
-            @test_throws ErrorException gs.hidden_state
-            @test_throws ErrorException gs.memory
+                lstm = LSTMCell(
+                    3 => 5; use_bias=false, train_state=false, train_memory=false)
+                ps, st = Lux.setup(rng, lstm) .|> device
+                ps = _ps
+                (y, carry), _ = Lux.apply(lstm, x, ps, st)
+                @test carry == _carry
+                l, back = Zygote.pullback(
+                    p -> sum(abs2, 0 .- sum(lstm(x, p, st)[1][1])), ps)
+                gs = back(one(l))[1]
+                @test_throws ErrorException gs.bias
+                @test_throws ErrorException gs.hidden_state
+                @test_throws ErrorException gs.memory
 
-            lstm = LSTMCell(3 => 5; use_bias=false, train_state=true, train_memory=false)
-            ps, st = Lux.setup(rng, lstm) .|> device
-            ps = merge(_ps, (hidden_state=ps.hidden_state,))
-            (y, carry), _ = Lux.apply(lstm, x, ps, st)
-            @test carry == _carry
-            l, back = Zygote.pullback(p -> sum(abs2, 0 .- sum(lstm(x, p, st)[1][1])), ps)
-            gs = back(one(l))[1]
-            @test_throws ErrorException gs.bias
-            @test !isnothing(gs.hidden_state)
-            @test_throws ErrorException gs.memory
+                lstm = LSTMCell(
+                    3 => 5; use_bias=false, train_state=true, train_memory=false)
+                ps, st = Lux.setup(rng, lstm) .|> device
+                ps = merge(_ps, (hidden_state=ps.hidden_state,))
+                (y, carry), _ = Lux.apply(lstm, x, ps, st)
+                @test carry == _carry
+                l, back = Zygote.pullback(
+                    p -> sum(abs2, 0 .- sum(lstm(x, p, st)[1][1])), ps)
+                gs = back(one(l))[1]
+                @test_throws ErrorException gs.bias
+                @test !isnothing(gs.hidden_state)
+                @test_throws ErrorException gs.memory
 
-            lstm = LSTMCell(3 => 5; use_bias=false, train_state=false, train_memory=true)
-            ps, st = Lux.setup(rng, lstm) .|> device
-            ps = merge(_ps, (memory=ps.memory,))
-            (y, carry), _ = Lux.apply(lstm, x, ps, st)
-            @test carry == _carry
-            l, back = Zygote.pullback(p -> sum(abs2, 0 .- sum(lstm(x, p, st)[1][1])), ps)
-            gs = back(one(l))[1]
-            @test_throws ErrorException gs.bias
-            @test_throws ErrorException gs.hidden_state
-            @test !isnothing(gs.memory)
+                lstm = LSTMCell(
+                    3 => 5; use_bias=false, train_state=false, train_memory=true)
+                ps, st = Lux.setup(rng, lstm) .|> device
+                ps = merge(_ps, (memory=ps.memory,))
+                (y, carry), _ = Lux.apply(lstm, x, ps, st)
+                @test carry == _carry
+                l, back = Zygote.pullback(
+                    p -> sum(abs2, 0 .- sum(lstm(x, p, st)[1][1])), ps)
+                gs = back(one(l))[1]
+                @test_throws ErrorException gs.bias
+                @test_throws ErrorException gs.hidden_state
+                @test !isnothing(gs.memory)
 
-            lstm = LSTMCell(3 => 5; use_bias=false, train_state=true, train_memory=true)
-            ps, st = Lux.setup(rng, lstm) .|> device
-            ps = merge(_ps, (hidden_state=ps.hidden_state, memory=ps.memory))
-            (y, carry), _ = Lux.apply(lstm, x, ps, st)
-            @test carry == _carry
-            l, back = Zygote.pullback(p -> sum(abs2, 0 .- sum(lstm(x, p, st)[1][1])), ps)
-            gs = back(one(l))[1]
-            @test_throws ErrorException gs.bias
-            @test !isnothing(gs.hidden_state)
-            @test !isnothing(gs.memory)
+                lstm = LSTMCell(3 => 5; use_bias=false, train_state=true, train_memory=true)
+                ps, st = Lux.setup(rng, lstm) .|> device
+                ps = merge(_ps, (hidden_state=ps.hidden_state, memory=ps.memory))
+                (y, carry), _ = Lux.apply(lstm, x, ps, st)
+                @test carry == _carry
+                l, back = Zygote.pullback(
+                    p -> sum(abs2, 0 .- sum(lstm(x, p, st)[1][1])), ps)
+                gs = back(one(l))[1]
+                @test_throws ErrorException gs.bias
+                @test !isnothing(gs.hidden_state)
+                @test !isnothing(gs.memory)
 
-            lstm = LSTMCell(3 => 5; use_bias=true, train_state=true, train_memory=true)
-            ps, st = Lux.setup(rng, lstm) .|> device
-            ps = merge(_ps, (bias=ps.bias, hidden_state=ps.hidden_state, memory=ps.memory))
-            (y, carry), _ = Lux.apply(lstm, x, ps, st)
-            l, back = Zygote.pullback(p -> sum(abs2, 0 .- sum(lstm(x, p, st)[1][1])), ps)
-            gs = back(one(l))[1]
-            @test !isnothing(gs.bias)
-            @test !isnothing(gs.hidden_state)
-            @test !isnothing(gs.memory)
+                lstm = LSTMCell(3 => 5; use_bias=true, train_state=true, train_memory=true)
+                ps, st = Lux.setup(rng, lstm) .|> device
+                ps = merge(
+                    _ps, (bias=ps.bias, hidden_state=ps.hidden_state, memory=ps.memory))
+                (y, carry), _ = Lux.apply(lstm, x, ps, st)
+                l, back = Zygote.pullback(
+                    p -> sum(abs2, 0 .- sum(lstm(x, p, st)[1][1])), ps)
+                gs = back(one(l))[1]
+                @test !isnothing(gs.bias)
+                @test !isnothing(gs.hidden_state)
+                @test !isnothing(gs.memory)
+            end
         end
     end
 end
@@ -150,59 +172,64 @@ end
             GRUCell(3 => 5; use_bias=false))
             __display(grucell)
             ps, st = Lux.setup(rng, grucell) .|> device
-            x = randn(rng, Float32, 3, 2) |> aType
-            (y, carry), st_ = Lux.apply(grucell, x, ps, st)
 
-            @jet grucell(x, ps, st)
-            @jet grucell((x, carry), ps, st)
+            for x_size in ((3, 2), (3,))
+                x = randn(rng, Float32, x_size...) |> aType
+                (y, carry), st_ = Lux.apply(grucell, x, ps, st)
 
-            function loss_loop_grucell(p)
-                (y, carry), st_ = grucell(x, p, st)
-                for i in 1:10
-                    (y, carry), st_ = grucell((x, carry), p, st_)
+                @jet grucell(x, ps, st)
+                @jet grucell((x, carry), ps, st)
+
+                function loss_loop_grucell(p)
+                    (y, carry), st_ = grucell(x, p, st)
+                    for i in 1:10
+                        (y, carry), st_ = grucell((x, carry), p, st_)
+                    end
+                    return sum(abs2, y)
                 end
-                return sum(abs2, y)
+
+                @eval @test_gradients $loss_loop_grucell $ps atol=1e-2 rtol=1e-2 gpu_testing=$ongpu
+
+                @test_throws ErrorException ps.train_state
             end
-
-            @eval @test_gradients $loss_loop_grucell $ps atol=1e-2 rtol=1e-2 gpu_testing=$ongpu
-
-            @test_throws ErrorException ps.train_state
         end
 
         @testset "Trainable hidden states" begin
-            x = randn(rng, Float32, 3, 2) |> aType
-            _gru = GRUCell(3 => 5; use_bias=false, train_state=false)
-            _ps, _st = Lux.setup(rng, _gru) .|> device
-            (_y, _carry), _ = Lux.apply(_gru, x, _ps, _st)
+            for x_size in ((3, 2), (3,))
+                x = randn(rng, Float32, x_size...) |> aType
+                _gru = GRUCell(3 => 5; use_bias=false, train_state=false)
+                _ps, _st = Lux.setup(rng, _gru) .|> device
+                (_y, _carry), _ = Lux.apply(_gru, x, _ps, _st)
 
-            gru = GRUCell(3 => 5; use_bias=false, train_state=false)
-            ps, st = Lux.setup(rng, gru) .|> device
-            ps = _ps
-            (y, carry), _ = Lux.apply(gru, x, ps, st)
-            @test carry == _carry
-            l, back = Zygote.pullback(p -> sum(abs2, 0 .- sum(gru(x, p, st)[1][1])), ps)
-            gs = back(one(l))[1]
-            @test_throws ErrorException gs.bias
-            @test_throws ErrorException gs.hidden_state
+                gru = GRUCell(3 => 5; use_bias=false, train_state=false)
+                ps, st = Lux.setup(rng, gru) .|> device
+                ps = _ps
+                (y, carry), _ = Lux.apply(gru, x, ps, st)
+                @test carry == _carry
+                l, back = Zygote.pullback(p -> sum(abs2, 0 .- sum(gru(x, p, st)[1][1])), ps)
+                gs = back(one(l))[1]
+                @test_throws ErrorException gs.bias
+                @test_throws ErrorException gs.hidden_state
 
-            gru = GRUCell(3 => 5; use_bias=false, train_state=true)
-            ps, st = Lux.setup(rng, gru) .|> device
-            ps = merge(_ps, (hidden_state=ps.hidden_state,))
-            (y, carry), _ = Lux.apply(gru, x, ps, st)
-            @test carry == _carry
-            l, back = Zygote.pullback(p -> sum(abs2, 0 .- sum(gru(x, p, st)[1][1])), ps)
-            gs = back(one(l))[1]
-            @test !isnothing(gs.hidden_state)
+                gru = GRUCell(3 => 5; use_bias=false, train_state=true)
+                ps, st = Lux.setup(rng, gru) .|> device
+                ps = merge(_ps, (hidden_state=ps.hidden_state,))
+                (y, carry), _ = Lux.apply(gru, x, ps, st)
+                @test carry == _carry
+                l, back = Zygote.pullback(p -> sum(abs2, 0 .- sum(gru(x, p, st)[1][1])), ps)
+                gs = back(one(l))[1]
+                @test !isnothing(gs.hidden_state)
 
-            gru = GRUCell(3 => 5; use_bias=true, train_state=true)
-            ps, st = Lux.setup(rng, gru) .|> device
-            ps = merge(
-                _ps, (bias_h=ps.bias_h, bias_i=ps.bias_i, hidden_state=ps.hidden_state))
-            (y, carry), _ = Lux.apply(gru, x, ps, st)
-            @test carry == _carry
-            l, back = Zygote.pullback(p -> sum(abs2, 0 .- sum(gru(x, p, st)[1][1])), ps)
-            gs = back(one(l))[1]
-            @test !isnothing(gs.hidden_state)
+                gru = GRUCell(3 => 5; use_bias=true, train_state=true)
+                ps, st = Lux.setup(rng, gru) .|> device
+                ps = merge(
+                    _ps, (bias_h=ps.bias_h, bias_i=ps.bias_i, hidden_state=ps.hidden_state))
+                (y, carry), _ = Lux.apply(gru, x, ps, st)
+                @test carry == _carry
+                l, back = Zygote.pullback(p -> sum(abs2, 0 .- sum(gru(x, p, st)[1][1])), ps)
+                gs = back(one(l))[1]
+                @test !isnothing(gs.hidden_state)
+            end
         end
     end
 end
@@ -218,30 +245,37 @@ end
             cell = _cell(3 => 5; use_bias, train_state)
             rnn = StatefulRecurrentCell(cell)
             __display(rnn)
-            x = randn(rng, Float32, 3, 2) |> aType
-            ps, st = Lux.setup(rng, rnn) .|> device
 
-            y, st_ = rnn(x, ps, st)
+            for x_size in ((3, 2), (3,))
+                x = randn(rng, Float32, x_size...) |> aType
+                ps, st = Lux.setup(rng, rnn) .|> device
 
-            @jet rnn(x, ps, st)
-            @jet rnn(x, ps, st_)
+                y, st_ = rnn(x, ps, st)
 
-            @test size(y) == (5, 2)
-            @test st.carry === nothing
-            @test st_.carry !== nothing
+                @jet rnn(x, ps, st)
+                @jet rnn(x, ps, st_)
 
-            st__ = Lux.update_state(st, :carry, nothing)
-            @test st__.carry === nothing
-
-            function loss_loop_rnn(p)
-                y, st_ = rnn(x, p, st)
-                for i in 1:10
-                    y, st_ = rnn(x, p, st_)
+                if length(x_size) == 2
+                    @test size(y) == (5, 2)
+                else
+                    @test size(y) == (5,)
                 end
-                return sum(abs2, y)
-            end
+                @test st.carry === nothing
+                @test st_.carry !== nothing
 
-            @eval @test_gradients $loss_loop_rnn $ps atol=1e-2 rtol=1e-2 gpu_testing=$ongpu
+                st__ = Lux.update_state(st, :carry, nothing)
+                @test st__.carry === nothing
+
+                function loss_loop_rnn(p)
+                    y, st_ = rnn(x, p, st)
+                    for i in 1:10
+                        y, st_ = rnn(x, p, st_)
+                    end
+                    return sum(abs2, y)
+                end
+
+                @eval @test_gradients $loss_loop_rnn $ps atol=1e-2 rtol=1e-2 gpu_testing=$ongpu
+            end
         end
     end
 end
@@ -283,7 +317,37 @@ end
                         @test length(y_) == 4
                         @test all(x -> size(x) == (5, 2), y_)
 
-                        if mode != "AMDGPU" && !(VERSION < v"1.9" && x isa AbstractVector)
+                        if mode != "AMDGPU"
+                            __f = p -> sum(first(rnn(x, p, st)))
+                            @eval @test_gradients $__f $ps atol=1e-2 rtol=1e-2 gpu_testing=$ongpu
+
+                            __f = p -> sum(Base.Fix1(sum, abs2), first(rnn_seq(x, p, st)))
+                            @eval @test_gradients $__f $ps atol=1e-2 rtol=1e-2 gpu_testing=$ongpu
+                        else
+                            # This is just added as a stub to remember about this broken test
+                            @test_broken 1 + 1 == 1
+                        end
+                    end
+
+                    ordering isa BatchLastIndex && continue
+
+                    # Batched Time Series without data batches
+                    @testset "typeof(x): $(typeof(x))" for x in (
+                        randn(rng, Float32, 3, 4) |> aType,
+                        Tuple(randn(rng, Float32, 3) for _ in 1:4) .|> aType,
+                        [randn(rng, Float32, 3) for _ in 1:4] .|> aType)
+                        ps, st = Lux.setup(rng, rnn) .|> device
+                        y, st_ = rnn(x, ps, st)
+                        y_, st__ = rnn_seq(x, ps, st)
+
+                        @jet rnn(x, ps, st)
+                        @jet rnn_seq(x, ps, st)
+
+                        @test size(y) == (5,)
+                        @test length(y_) == 4
+                        @test all(x -> size(x) == (5,), y_)
+
+                        if mode != "AMDGPU"
                             __f = p -> sum(first(rnn(x, p, st)))
                             @eval @test_gradients $__f $ps atol=1e-2 rtol=1e-2 gpu_testing=$ongpu
 

